shared-bindings/gamepad/__init__.c

```diff
@@ -0,0 +1,101 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2016 Radomir Dopieralski for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include "py/obj.h"
+#include "py/runtime.h"
+#include "py/mphal.h"
+#include "shared-module/gamepad/__init__.h"
+
+
+gamepad_obj_t* gamepad_singleton = NULL;
+
+
+
+//| :mod:`gamepad` --- Button handling
+//  ==================================
+//|
+//| .. module:: gamepad
+//|   :synopsis: Button handling
+//|   :platform: SAMD21
+//|
+
+//| ..method:: GamePad.get_pressed()
```

Comment:
Move this below the class, then indent it and it'll be included in the class automatically. Here is an example: https://github.com/adafruit/circuitpython/blob/master/shared-bindings/touchio/TouchIn.c#L54

I also think `pressed` should be a property rather than a method because its state.

Reply:
> I also think pressed should be a property rather than a method because its state.

I would make it a property, if not for the side effect, which is clearing the current state. The way this works, it accumulates button presses, until you call "get_pressed", which gives you what has been pressed since the last call, and clears the buffer. Making it a property would lead to surprising, unexpected behaviors where the status is cleared more than once, etc.

I made the status clearing functionality more explicit in the description.

Reply:
Also note, that it is important that the status is retrieved and cleared in one atomic operation, and not first retrieved, and then cleared separately — the latter risks missing some button presses.

Reply:
Ok! Looks good.

Another critical point in my mind is that it includes past presses. The state module works better when there are no side effects and it only is actively pressed buttons. Thanks for the push back!